src/Bicep.Core/TypeSystem/UnassignableType.cs

```diff
@@ -6,29 +6,34 @@
 
 namespace Bicep.Core.TypeSystem
 {
-    public class ErrorTypeSymbol : TypeSymbol
+    public class UnassignableTypeSymbol : TypeSymbol
     {
         private const string ErrorTypeName = "error";
         private readonly ImmutableArray<ErrorDiagnostic> errors;
 
-        public ErrorTypeSymbol(ErrorDiagnostic error)
+        public UnassignableTypeSymbol(string name)
+            : base(name)
+        {
+        }
+
+        public UnassignableTypeSymbol(ErrorDiagnostic error)
             : base(ErrorTypeName)
         {
             this.errors = ImmutableArray.Create<ErrorDiagnostic>(error);
         }
 
-        public ErrorTypeSymbol(IEnumerable<ErrorDiagnostic> errors)
+        public UnassignableTypeSymbol(IEnumerable<ErrorDiagnostic> errors)
             : base(ErrorTypeName)
         {
             this.errors = errors.ToImmutableArray();
         }
 
         public override IEnumerable<ErrorDiagnostic> GetDiagnostics()
         {
-            return this.errors;
+            return this.errors.IsDefaultOrEmpty ? ImmutableArray<ErrorDiagnostic>.Empty : this.errors;
         }
 
-        public override TypeKind TypeKind => TypeKind.Error;
+        public override TypeKind TypeKind => this.Name.Equals(ErrorTypeName) ? TypeKind.Error : TypeKind.Unassignable;
```

Comment:
Can we pass this in explicitly through the constructor rather than depending on the name?

Reply:
Updated the constructor overload that receives a name.